### PR TITLE
Add support for nested list arrays from parquet to arrow arrays (#993)

### LINKED ISSUE #720: Support for nested data types
While reading a parquet file using arrow rust, I ran into an issue 
```
ArrowError("reading List(GroupType ... into arrow not supported yet")
```

Any timeline for supporting nested structures would be of great help.

Thanks!

### LINKED ISSUE #993: Support Reading Nested List Arrays
**Describe the bug**

When trying to read parquet files with deeply nested fields we end up with the following error:

```
Parquet reader thread terminated due to error: ParquetError(ArrowError("reading List(GroupType {
...
}) into arrow not supported yet"))
```

**To Reproduce**

This is easily visible from the code found at [array_reader.rs#L1516-L1522](https://github.com/apache/arrow-rs/blob/master/parquet/src/arrow/array_reader.rs#L1516-L1522)

**Expected behavior**

To have support for reading nested parquet files into arrow.

**Additional context**

This issue, in my particular case, has been hidden under [#1383](https://github.com/apache/arrow-datafusion/issues/1383).

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #993.

Closes #720 

# Rationale for this change
 
Adds support for reading nested lists from parquet.

# What changes are included in this PR?

Reworks the ListArrayReader to handle nested repetition levels.

Some drive by cleanup of ArrayReaderBuilder to error on encountering non-spec-compliant data

# Are there any user-facing changes?

No

### CODE REVIEW

File: parquet/src/arrow/array_reader/list_array.rs [hunk 1]
```diff
@@ -212,7 +196,7 @@ mod tests {
         let item_array_reader = InMemoryArrayReader::new(
             ArrowType::Int32,
             array,
-            Some(vec![3, 2, 3, 0, 3, 3]),
+            Some(vec![2, 1, 2, 0, 2, 2]),
```
Comment (tustvold) [inline parquet/src/arrow/array_reader/list_array.rs:199]: This is just a drive-by fix, it isn't required to get the test to pass, but was technically wrong before

File: parquet/src/arrow/array_reader/list_array.rs [hunk 2]
```diff
@@ -88,97 +84,85 @@ impl<OffsetSize: OffsetSizeTrait> ArrayReader for ListArrayReader<OffsetSize> {
...
-        // If a Parquet schema's only leaf is the list, then n = 0.
-
-        // If the list index is at empty definition, the child slot is null
-        let non_null_list_indices =
-            def_levels.iter().enumerate().filter_map(|(index, def)| {
-                (*def > self.list_empty_def_level).then(|| index as u32)
-            });
-        let indices = UInt32Array::from_iter_values(non_null_list_indices);
-        let batch_values =
-            arrow::compute::take(&*next_batch_array.clone(), &indices, None)?;
```
Comment (tustvold) [inline parquet/src/arrow/array_reader/list_array.rs:123]: We no longer do this, which is likely better for performance, but comes with the caveat that nulls take up more space than before.

File: parquet/src/arrow/array_reader/builder.rs [hunk 3]
```diff
@@ -208,42 +191,61 @@ impl<'a> TypeVisitor<Option<Box<dyn ArrayReader>>, &'a ArrayReaderBuilderContext
...
-            .ok_or_else(|| ArrowError("Map entry must have a key".to_string()))?;
-        let map_value = map_key_value
-            .get_fields()
-            .get(1)
-            .ok_or_else(|| ArrowError("Map entry must have a value".to_string()))?;
-
-        let key_reader = {
-            let mut key_context = new_context.clone();
-            key_context.def_level += 1;
-            key_context.path.append(vec![map_key.name().to_string()]);
```
Comment (tustvold) [inline parquet/src/arrow/array_reader/builder.rs:237]: This is a drive-by fix, the context is the context of the parent, not the value being dispatched. This would result in map_key appearing twice

File: parquet/src/arrow/array_reader/builder.rs [hunk 4]
```diff
@@ -295,96 +297,80 @@ impl<'a> TypeVisitor<Option<Box<dyn ArrayReader>>, &'a ArrayReaderBuilderContext
             .first()
             .ok_or_else(|| ArrowError("List field must have a child.".to_string()))?
             .clone();
-        let mut new_context = context.clone();
 
+        // If the list can contain nulls
```
Comment (tustvold) [inline parquet/src/arrow/array_reader/builder.rs:301]: I recommend reading this with whitespace disabled https://github.com/apache/arrow-rs/pull/1588/files?w=1

File: parquet/src/arrow/array_reader/list_array.rs [hunk 5]
```diff
@@ -193,122 +239,267 @@ mod tests {
...
+
+    fn to_offsets<OffsetSize: OffsetSizeTrait>(values: Vec<usize>) -> Buffer {
+        Buffer::from_iter(
+            values
+                .into_iter()
+                .map(|x| OffsetSize::from_usize(x).unwrap()),
+        )
+    }
+
+    fn test_nested_list<OffsetSize: OffsetSizeTrait>() {
```
Comment (tustvold) [inline parquet/src/arrow/array_reader/list_array.rs:274]: I think this wins the prize of the most :exploding_head: test I've ever written

File: parquet/src/arrow/array_reader/builder.rs [hunk 6]
```diff
@@ -129,9 +129,10 @@ impl<'a> TypeVisitor<Option<Box<dyn ArrayReader>>, &'a ArrayReaderBuilderContext
 
             let null_mask_only = match cur_type.get_basic_info().repetition() {
                 Repetition::REPEATED => {
-                    new_context.def_level += 1;
-                    new_context.rep_level += 1;
-                    false
+                    return Err(ArrowError(format!(
```
Comment (tustvold) [inline parquet/src/arrow/array_reader/builder.rs:132]: This just moves the error to earlier, there is no point continuing if this is not supported anyway :grin:

File: parquet/src/arrow/array_reader/builder.rs [hunk 7]
```diff
@@ -295,96 +297,80 @@ impl<'a> TypeVisitor<Option<Box<dyn ArrayReader>>, &'a ArrayReaderBuilderContext
...
 
-        let reader = self.dispatch(item_type.clone(), &new_context);
-        if let Ok(Some(item_reader)) = reader {
-            let item_reader_type = item_reader.get_data_type().clone();
-
-            match item_reader_type {
-                ArrowType::List(_)
-                | ArrowType::FixedSizeList(_, _)
-                | ArrowType::Dictionary(_, _) => Err(ArrowError(format!(
-                    "reading List({:?}) into arrow not supported yet",
```
Comment (tustvold) [inline parquet/src/arrow/array_reader/builder.rs:331]: This is the error that would previously be returned on a nested list

File: parquet/src/schema/visitor.rs [hunk 8]
```diff
@@ -27,17 +27,30 @@ pub trait TypeVisitor<R, C> {
...
     /// Default implementation when visiting a list.
     ///
-    /// It checks list type definition and calls `visit_list_with_item` with extracted
+    /// It checks list type definition and calls [`Self::visit_list_with_item`] with extracted
     /// item type.
     ///
     /// To fully understand this algorithm, please refer to
     /// [parquet doc](https://github.com/apache/parquet-format/blob/master/LogicalTypes.md).
+    ///
+    /// For example, a standard list type looks like:
```
Comment (alamb) [inline parquet/src/schema/visitor.rs:36]: ❤️

File: parquet/src/arrow/array_reader/list_array.rs [hunk 9]
```diff
@@ -35,12 +35,12 @@ pub struct ListArrayReader<OffsetSize: OffsetSizeTrait> {
     item_reader: Box<dyn ArrayReader>,
     data_type: ArrowType,
     item_type: ArrowType,
-    list_def_level: i16,
-    list_rep_level: i16,
-    list_empty_def_level: i16,
-    list_null_def_level: i16,
-    def_level_buffer: Option<Buffer>,
-    rep_level_buffer: Option<Buffer>,
+    // The definition level at which this list is not null
```
Comment (alamb) [inline parquet/src/arrow/array_reader/list_array.rs:38]:
```suggestion
    /// The definition level at which this list is not null
```

File: parquet/src/arrow/array_reader/list_array.rs [hunk 10]
```diff
@@ -35,12 +35,12 @@ pub struct ListArrayReader<OffsetSize: OffsetSizeTrait> {
...
+    def_level: i16,
+    // The repetition level that corresponds to a new value in this array
```
Comment (alamb) [inline parquet/src/arrow/array_reader/list_array.rs:40]:
```suggestion
    /// The repetition level that corresponds to a new value in this array
```

File: parquet/src/arrow/array_reader/list_array.rs [hunk 11]
```diff
@@ -35,12 +35,12 @@ pub struct ListArrayReader<OffsetSize: OffsetSizeTrait> {
...
+    rep_level: i16,
+    // If this list is nullable
```
Comment (alamb) [inline parquet/src/arrow/array_reader/list_array.rs:42]:
```suggestion
    /// If this list is nullable
```

File: parquet/src/arrow/array_reader/list_array.rs [hunk 12]
```diff
@@ -193,122 +245,267 @@ mod tests {
...
+        // [[[1, null], null, [4], []],[], [[7]], [[]], [[11]]]
+        let offsets = to_offsets::<OffsetSize>(vec![0, 4, 4, 5, 6, 7]);
+        let l2 = ArrayDataBuilder::new(l2_type.clone())
+            .len(5)
+            .add_buffer(offsets)
+            .add_child_data(l3)
+            .build()
+            .unwrap();
+
+        let offsets = to_offsets::<OffsetSize>(vec![0, 4, 4, 4, 5]);
```
Comment (alamb) [inline parquet/src/arrow/array_reader/list_array.rs:330]: Would it be valuable to have a sublist that has more than 1 item? 

It appears that this test only has `null`, `[]` or a single element `[x]` list. I may be misunderstanding the intent of the test or what is possible in nested lists

File: parquet/src/arrow/array_reader/list_array.rs [hunk 13]
```diff
@@ -193,122 +245,267 @@ mod tests {
...
+        let mut l1 = ListArrayReader::<OffsetSize>::new(
+            Box::new(l2),
+            l1_type,
+            l1_item_type,
+            2,
+            1,
+            true,
+        );
+
+        let actual = l1.next_batch(1024).unwrap();
```
Comment (alamb) [inline parquet/src/arrow/array_reader/list_array.rs:389]: Would it also make sense to test with `l1.next_batch(2)` (aka something that doesn't decode the entire array in one go?)

**Review by alamb — APPROVED** [review]
I can't say I am a deep expert in this area or this part of the codebase, but I did give the PR a careful "software engineering" review, with careful study of the tests. LGTM 👍 

FWIW whitespace blind diff https://github.com/apache/arrow-rs/pull/1588/files?w=1 helped in my review

cc @nevi-me  and @sunchao  in case you would like to review

File: parquet/src/arrow/array_reader/list_array.rs (hunk 12 above)
Comment (tustvold) [reply parquet/src/arrow/array_reader/list_array.rs:330]: There is a `[1, null]`, can probably add more though

File: parquet/src/arrow/array_reader/list_array.rs (hunk 13 above)
Comment (tustvold) [reply parquet/src/arrow/array_reader/list_array.rs]: Done in https://github.com/apache/arrow-rs/pull/1588/commits/05c231190aef2c9054b9ebb7fbb59f4a1daf129f, required adding functionality to InMemoryArrayReader to actually respect it

File: parquet/src/arrow/array_reader/list_array.rs (hunk 12 above)
Comment (tustvold) [reply parquet/src/arrow/array_reader/list_array.rs]: Done in https://github.com/apache/arrow-rs/pull/1588/commits/05c231190aef2c9054b9ebb7fbb59f4a1daf129f

File: parquet/src/arrow/array_reader.rs [hunk 14]
```diff
@@ -1532,15 +1532,15 @@ mod tests {
             ArrowType::Int32,
             array_1.clone(),
             Some(vec![0, 1, 2, 3, 1]),
-            Some(vec![1, 1, 1, 1, 1]),
+            Some(vec![0, 1, 1, 1, 1]),
```
Comment (tustvold) [inline parquet/src/arrow/array_reader.rs:1535]: Drive by fix, this set of levels is impossible, as the first rep level must be 0. Nothing cares as this test doesn't actually decode the implied list, but :shrug:

File: parquet/src/arrow/array_reader/list_array.rs [hunk 15]
```diff
@@ -35,12 +35,12 @@ pub struct ListArrayReader<OffsetSize: OffsetSizeTrait> {
...
-    list_empty_def_level: i16,
-    list_null_def_level: i16,
-    def_level_buffer: Option<Buffer>,
-    rep_level_buffer: Option<Buffer>,
+    /// The definition level at which this list is not null
+    def_level: i16,
+    /// The repetition level that corresponds to a new value in this array
+    rep_level: i16,
+    /// If this list is nullable
+    nullable: bool,
```
Comment (viirya) [inline parquet/src/arrow/array_reader/list_array.rs:43]: The comment looks clear. 👍

File: parquet/src/schema/visitor.rs (hunk 8 above)
Comment (viirya) [reply parquet/src/schema/visitor.rs:36]: nice document addition 👍

**Review by viirya — APPROVED** [review]
Looked at the changes, especially on `ListArrayReader`. I'm not Parquet expert too, but the change looks reasonable when looking at with the added comments. 👍
